src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,380 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Linq;
+
+namespace Microsoft.Experimental.Collections
+{
+    /// <summary>
+    /// DictionarySlim<TKey, TValue> is similar to Dictionary<TKey, TValue> but optimized in three ways:
+    /// 1) It allows access to the value by ref.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer(assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        const int DefaultPrimeSize = 3;
+        // 1-based index into _entries; 0 means empty
+        private int[] _buckets;
+        private Entry[] _entries;
+        // 0-based index into _entries of head of free chain: -1 means empty
+        private int _freeList = -1;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            // 0-based index of next entry in chain: -1 means empty
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = new int[DefaultPrimeSize];
```

Comment:
@AnthonyLloyd currently an empty DictionarySlim is a lot less 'slim' than an empty Dictionary, which is unfortunate. if you start the arrays as Array.Empty, there should be no perf impact, I think? The existing resize check would grow them on the first add. Thoughts?

Reply:
There a division by zero from the modulo. I can't see how to stop this with a costly check at the moment.

Reply:
Ah of course. Hmm. Dictionary just has `if (buckets != null)`

What about this trick: we initialize with a static readonly array of size 1. The value of the single entry in this bucket will be 0, so "reads" of the dictionary will see it as empty without any extra checks. In the write codepath, we change
```
                    if (Count == entries.Length)
                    {
                        entries = Resize();
```
to
```
                    if (Count == entries.Length || entries.Length == 1)
                    {
                        entries = Resize();
```
Which will cause the dummy array to be replaced with a real one of size 3.

I didn't think this through carefully -- would it work? If so it seems the same trick could be used in Dictionary itself.

Reply:
Good one. It works and doesn't effect perf.

Reply:
Added